sched/semaphore/spinlock.c

```diff
@@ -446,4 +446,225 @@ void spin_clrbit(FAR volatile cpu_set_t *set, unsigned int cpu,
 }
 #endif
 
+#ifdef CONFIG_RW_SPINLOCK
+
+/****************************************************************************
+ * Name: read_lock
+ *
+ * Description:
+ *   If this task does not already hold the spinlock, then loop until the
+ *   spinlock is successfully locked.
+ *
+ *   This implementation is non-reentrant and set a bit of lock.
+ *
+ *  The priority of reader is higher than writter if a reader hold the
+ *  lock, a new reader can get its lock but writer can't get this lock.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   None.  When the function returns, the spinlock was successfully locked
+ *   by this CPU.
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void read_lock(FAR volatile rwlock_t *lock)
+{
+  int old;
```

Comment:
In order to avoid compiler issue on Mac sim-02
```
Error: semaphore/spinlock.c:630:45: error: incompatible pointer types passing 'rwlock_t *' (aka '_Atomic(int) *') to parameter of type 'int *' [-Werror,-Wincompatible-pointer-types]
[1446](https://github.com/apache/nuttx/actions/runs/6485231606/job/17611025912#step:7:1447)
  if (!atomic_compare_exchange_strong(lock, &zero, RW_SP_WRITE_LOCKED))
[1447](https://github.com/apache/nuttx/actions/runs/6485231606/job/17611025912#step:7:1448)
                                            ^~~~~
```

Reply:
@xiaoxiang781216 Is it a good idea to splite two commit on this PR?
One is the implementation of read-write spinlock and the other is for resolving Mac sim-02 compiler issue.
Maybe we can revert the second commit if resolving this compiler issue of Mac sim-02.